data/basal.go

```diff
@@ -0,0 +1,166 @@
+
+	levemirInsulin, lantusInsulin = "levemir", "lantus"
+
+	rateValidationLowerLimit     = 0.0
+	durationValidationLowerLimit = 0
+	valueValidationLowerLimit    = 0
+	validationGreaterThanInt     = "Must be greater than %d"
+)
+
+var (
+	allowedDeliveryTypes = map[string]string{injectedDelivery: injectedDelivery, scheduledDelivery: scheduledDelivery, suspendDelivery: suspendDelivery, tempDelivery: tempDelivery}
+	allowedInsulins      = map[string]string{levemirInsulin: levemirInsulin, lantusInsulin: lantusInsulin}
+
+	basalFailureReasons = validate.ErrorReasons{
+		deliveryTypeTag: fmt.Sprintf("Must be one of %s,%s,%s,%s", injectedDelivery, scheduledDelivery, suspendDelivery, tempDelivery),
+		rateTag:         fmt.Sprintf("Must be greater than %f", rateValidationLowerLimit),
+		durationTag:     fmt.Sprintf(validationGreaterThanInt, durationValidationLowerLimit),
+		valueTag:        fmt.Sprintf(validationGreaterThanInt, valueValidationLowerLimit),
+		insulinTag:      fmt.Sprintf("Must be one of %s,%s", levemirInsulin, lantusInsulin),
+	}
+)
+
+//BuildBasal will build a Basal record
+func BuildBasal(obj map[string]interface{}) (*Basal, *Error) {
+
+	base, errs := BuildBase(obj)
+	cast := NewCaster(errs)
+
+	basal := &Basal{
+		DeliveryType: cast.ToString(deliveryTypeField, obj[deliveryTypeField]),
+		ScheduleName: cast.ToString(scheduleNameField, obj[scheduleNameField]),
+		Base:         base,
+	}
+
+	if obj[rateField] != nil {
+		basal.Rate = cast.ToFloat64(rateField, obj[rateField])
+	}
+	if obj[durationField] != nil {
+		basal.Duration = cast.ToInt(durationField, obj[durationField])
+	}
+	if obj[insulinField] != nil {
+		basal.Insulin = cast.ToString(insulinField, obj[insulinField])
+	}
+
+	if validationErrors := validator.Struct(basal); len(validationErrors) > 0 {
+		errs.AppendError(validationErrors.GetError(basalFailureReasons))
+	}
+
+	if errs.IsEmpty() {
+		return basal, nil
+	}
+	return basal, errs
+}
+
+func BasalRateValidator(v *valid.Validate, topStruct reflect.Value, currentStructOrField reflect.Value, field reflect.Value, fieldType reflect.Type, fieldKind reflect.Kind, param string) bool {
+	if rate, ok := field.Interface().(float64); ok {
+		if rate > rateValidationLowerLimit {
+			return true
+		}
+	}
+	return false
+}
+
+func BasalDurationValidator(v *valid.Validate, topStruct reflect.Value, currentStructOrField reflect.Value, field reflect.Value, fieldType reflect.Type, fieldKind reflect.Kind, param string) bool {
+	if duration, ok := field.Interface().(int); ok {
+		if duration > durationValidationLowerLimit {
+			return true
+		}
+	}
+	return false
+}
+
+func BasalDeliveryTypeValidator(v *valid.Validate, topStruct reflect.Value, currentStructOrField reflect.Value, field reflect.Value, fieldType reflect.Type, fieldKind reflect.Kind, param string) bool {
+	if deliveryType, ok := field.Interface().(string); ok {
+		if _, ok = allowedDeliveryTypes[strings.ToLower(deliveryType)]; ok {
```

Comment:
I wonder if it makes sense to normalize these types of strings earlier in the processing so we know it will always be lowercase?  Perhaps as part of the Caster?  For example, `ToLowercaseString`?